ITE/4/README.adoc

```diff
@@ -0,0 +1,493 @@
+instead have a generic URI which can be resolved to some external resource that
+is part of the supply chain. Such a generic URI MUST have a fixed structure and
+MUST contain all the information necessary for the verifier to resolve if any
+part of the verification workflow requires it.
+
+This ITE proposes the following structure for the URI.
+
+`<scheme>:<hier-part>`
+
+This structure is derived from
+link:https://tools.ietf.org/html/rfc3986[RFC 3986]. The `scheme` defines the
+token or identifier for the type of resource the URI refers to. It can also
+contain other relevant information that indicates how the resource can be
+accessed. The `hier-part` identifies the path or location of the resource.
+
+A compliant in-toto system must be capable of generating link metadata with
+artifacts that have generic URI references. Therefore, this system must also be
+capable of resolving this URI and calculating cryptographic hashes of these
+artifacts. It's important to note that the contents of the resource are hashed,
+as opposed to the URI itself. This ITE doesn't specify how the contents are
+resolved and hashed by the system generating the metadata for the relevant
+abstract entities. All it specifies is that the resolution and the URI itself
+must be consistent in all metadata.
+
+Consider an example of a GitHub specific resource - a pull request or an issue.
+
+`github+https://github.com/in-toto/in-toto/pull/250`
+
+This can also instead use GitHub's API specific URI to refer to the same data.
+
+`github+https://api.github.com/repos/in-toto/in-toto/pulls/250`
+
+Both the above URIs refer to the same abstract entity - pull request 250 to
+in-toto's reference implementation. However, taken directly, they correspond to
+different data. Therefore, the implementing system must resolve the URI and thus
+the artifact in a consistent manner. We discuss this and other sources of
+non-determinism in the Appendix.
+
+An example of the product field for link metadata with a generic URI is as
+follows:
+
+```
+"products": {
+    "github+https://api.github.com/repos/in-toto/in-toto/pulls/250": {
+        "sha256": "7ee0617..."
+    },
+    "...": ...
+}
+```
+
+Similarly, a generic URI can be used to refer to other abstract entities like an
+SPDX element. The URI can be used to directly fetch the corresponding documents
+from a specified location. For in-toto metadata that are embedded within SPDX
+documents, the URI can also be used to refer to other entities elsewhere in the
+document. Resolving such URIs that point to local entities elsewhere in the SPDX
+document is an implementation level detail - but it's clear that the URI must
+contain all the information necessary to identify the entity in that context.
+
+It is important to note that this ITE doesn't mandate any specific generic URI
+tokens that all conforming implementations MUST support. Instead, it merely
+specifies what any generic URI must look like, and what rules the system MUST
```

Comment:
Can you please review the document for uniform case in [key words](https://www.ietf.org/rfc/rfc2119.txt).

Reply:
As I understand it, it doesn't apply to every occurrence of one of the words, correct?

Reply:
Correct. I don't know the exact rules though. I just noticed in above two lines that there are two *MUST* and one *must* and I couldn't see the pattern. If you chose different case for a reason then it is fine. :)

Reply:
No, I reviewed it again, I have the changes locally. During this review, I felt it didn't apply for every occurrence.

Reply:
Hope this looks better @lukpueh :)